lib/services/access_request.go

```diff
@@ -442,25 +458,112 @@ func newRequestValidator(traits map[string][]string, opts ...ValidateRequestOpti
 		m.Annotations.Allow = make(map[string][]string)
 		m.Annotations.Deny = make(map[string][]string)
 	}
-	return m
+
+	// load all statically assigned roles for the user and
+	// use them to build our validation state.
+	for _, roleName := range m.user.GetRoles() {
+		role, err := m.getter.GetRole(roleName)
+		if err != nil {
+			return RequestValidator{}, trace.Wrap(err)
+		}
+		if err := m.push(role); err != nil {
+			return RequestValidator{}, trace.Wrap(err)
+		}
+	}
+	return m, nil
+}
+
+// Validate validates an access request against
+func (m *RequestValidator) Validate(req AccessRequest) error {
+	if m.user.GetName() != req.GetUser() {
+		return trace.BadParameter("reqyest validator configured for different user (this is a bug)")
+	}
+
+	if m.requireReason && req.GetRequestReason() == "" {
+		return trace.BadParameter("request reason must be specified (required by static role configuration)")
+	}
+
+	// check for "wildcard request" (`roles=*`).  wildcard requests
+	// need to be expanded into a list consisting of all existing roles
+	// that the user does not hold and is allowed to request.
+	if r := req.GetRoles(); len(r) == 1 && r[0] == Wildcard {
+
+		if !req.GetState().IsPending() {
+			// expansion is only permitted in pending requests.  once resolved,
+			// a request's role list must be immutable.
+			return trace.BadParameter("wildcard requests are not permitted in state %s", req.GetState())
+		}
+
+		if !m.opts.expandRoles {
+			// teleport always validates new incoming pending access requests
+			// with ExpandRoles(true). after that, it should be impossible to
+			// add new values to the role list.
+			return trace.BadParameter("unexpected wildcard request (this is a bug)")
+		}
+
+		requestable, err := m.GetRequestableRoles()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if len(requestable) == 0 {
+			return trace.BadParameter("no requestable roles, please verify static RBAC configuration")
+		}
+		req.SetRoles(requestable)
+	}
+
+	// verify that all requested roles are permissible
+	for _, roleName := range req.GetRoles() {
+		if !m.CanRequestRole(roleName) {
+			return trace.BadParameter("user %q can not request role %q", req.GetUser(), roleName)
+		}
+	}
+
+	if m.opts.annotate {
+		// incoming requests must have system annotations attached before
+		// before being inserted into the backend. this is how the
+		// RBAC system propagates sideband information to plugins.
+		req.SetSystemAnnotations(m.SystemAnnotations())
+	}
+	return nil
+}
+
+// GetRequestableRoles gets the list of all existent roles which the user is
+// able to request.  This operation is expensive since it loads all existent
+// roles in order to determine the role list.  Prefer calling CanRequestRole
+// when checking againt a known role list.
+func (m *RequestValidator) GetRequestableRoles() ([]string, error) {
+	allRoles, err := m.getter.GetRoles()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var expanded []string
```

Comment:
Just curious.... If no roles, nil is returned and while it is technically correct it does get marshalled into null instead of an empty slice. It's not a big deal but just wondered if there was any real benefit between preferring nil slice `var expanded []string` over `expanded := make([]string, 0)`? I also wondered if it was better to return an empty slice, b/c description implies we get a list? Although len(nil slice) is zero... so maybe it does make sense to return a nil...

Reply:
The type definition includes the `omitempty` tag, so `requestable_roles` should not be present at all in the final object if the slice is `nil` or empty.  This is always my preference since it keeps objects less noisy by only including fields with meaningful values.